### ISSUE TITLE
Ollama (or other OpenAI API compatible local LLMs) support

### ISSUE THREAD

**Issue by micheleoletti** [getcursor/cursor#1380]
Knowing that Ollama server supports OpenAI API([https://ollama.com/blog/openai-compatibility](https://ollama.com/blog/openai-compatibility)), the goal is to **point Cursor to query the local Ollama server**.

My setup is pretty simple:
- Macbook Air M1
- Ollama running locally + llama2 

I added a `llama2` model, set `"ollama"` as API key(not used but needed apparently), and overridden the base URL to point to localhost.
<img width="503" alt="image" src="https://github.com/getcursor/cursor/assets/61937589/25e6eecf-b800-4e35-a95a-4a1f4295b229">

But it does not work:
<img width="533" alt="image" src="https://github.com/getcursor/cursor/assets/61937589/f4577523-2ee3-483d-be38-b4f85a8d45f4">

If I try to verify the API key it seems like it cannot reach localhost:
<img width="534" alt="image" src="https://github.com/getcursor/cursor/assets/61937589/867793e6-f143-4125-a0a0-cd754dcf8f83">

**But if I try the provided test snippet in the terminal, it works correctly:**
<img width="1438" alt="image" src="https://github.com/getcursor/cursor/assets/61937589/9a54ba4a-d382-44d4-b8a2-2b88d95ac745">

So it seems like **Cursor internal service is not able to perform the fetch to localhost**.

Is there something conceptually wrong with my plan and its implementation?
Did anybody manage to make this configuration work?

**Comment by tosh** [comment]
I also ran into this just now, is there a way to make it work?

**Comment by andreaspoldi** [comment]
+1

**Comment by TazzerMAN** [comment]
Same here :/

**Comment by griff12** [comment]
Cursor sends your request through their own servers and then to the baseurl that you specify. So yes this can work, but you need to use a public facing IP and handle port forwarding.

**Comment by kcolemangt** [comment]
Ollama Phi3 in the `⌘ K` 🚀

Takes some effort but it's fast and works well!

![cursor-proxy](https://github.com/getcursor/cursor/assets/20099734/7aca2faa-77fb-404c-8ce5-c0319ea24d1a)

**Comment by micheleoletti** [comment]
@kcolemangt looks amazing! what did you put in the settings to do that?
I can't get it to work 🤔

**Comment by kcolemangt** [comment]
Had to write a custom router and point Cursor’s `OpenAI Base URL` to that. Thinking of releasing it—want to try?

**Comment by micheleoletti** [comment]
I see... yeah I'd be interested in trying that out!
On what hardware are you running ollama by the way?

**Comment by kcolemangt** [comment]
released [llm-router](https://github.com/kcolemangt/llm-router). lmk how it works for you @micheleoletti

**Comment by yeralin** [comment]
Cursor does not like `127.0.0.1`

You'd need to use ngrok around your localhost: `ngrok http 11434 --host-header="localhost:11434"`

Put given address under "Override OpenAI Base URL" in Cursor: `https://<your-generated-link>.ngrok-free.app/v1`

Then it starts working.

**Comment by flight505** [comment]
there is the repo https://github.com/ryoppippi/proxy-worker-for-cursor-with-ollama which is an option (but not for offline use). The readme also points out there is direct communication to cursor server, which is likely why cursor haven't enabled an easy to use option for true local LLM.

**Comment by axsaucedo** [comment]
+1

I'm getting an error when using a custom URL even if the command works on the terminal

**Comment by prakashchokalingam** [comment]
+1

**Comment by hgoona** [comment]
> released [llm-router](https://github.com/kcolemangt/llm-router). lmk how it works for you @micheleoletti

Does this work on Windows? I'm wondering how to install?

**Comment by henry2man** [comment]
I've successfully configured Curxy for this mission https://github.com/ryoppippi/curxy

**Comment by hgoona** [comment]
> I've successfully configured Curxy for this mission https://github.com/ryoppippi/curxy

Hi @henry2man can you please elaborate on how you got it to work on Windows? 
I'm trying to use it with Groq Cloud and also Ollama (locally)

**Comment by henry2man** [comment]
> Hi @henry2man can you please elaborate on how you got it to work on Windows?
> I'm trying to use it with Groq Cloud and also Ollama (locally

Hi! I actually don’t use Windows, so I can’t share direct experience with that. However, I did notice that to get it working with CURXY and Ollama on MacOS, it’s crucial to enable the OpenAI custom API Key. Once that's done, you can override the OpenAI Base URL by adding a valid API key. Said that, it's a matter of time follow CURXY Readme instructions...

**Comment by Mateleo** [comment]
I did it for Windows using Ollama:

🚨 **Note:** Localhost is not working at the moment, so you’ll need to use a tunneling method. For this example, I used Ngrok.

1. **Install Ngrok**, register on the website, and validate your auth token [here](https://dashboard.ngrok.com/get-started/your-authtoken).  
2. Open a shell and use this command (from the Ollama doc):  
   `ngrok http 11434 --host-header="localhost:11434"`  
3. Use the forwarding link in **"Override OpenAI Base URL"**.

🔑 **We need an OpenAI API key to force Cursor to use our custom URL**:  
1. Generate a fake key [here](https://perchance.org/openaikeygen) and copy-paste it in the **"OpenAI API Key"** section.

Finally, disable all other models like so and add your custom model. Voilà! 🎉
![image](https://github.com/user-attachments/assets/4380622a-f754-4016-b0b4-533c78528b3e)

**Comment by hgoona** [comment]
Thanks @Mateleo! I'll give this a shot! BTW - when running Ollama with Cursor, does the "Cursor Prediction" work with custom models?

This feature: 
![image](https://github.com/user-attachments/assets/0bf406c3-5f28-4712-a217-683dbaa17bd1)

**Comment by Mateleo** [comment]
@hgoona Yes, and yet it's special... All the CTRL+K and Chat functions use the model, but I haven't noticed any calls to ollama for the 4 Tab features (they do work though). And yet in my settings on the site, I have no calls to external APIs gpt-4o or other.

**Comment by hhannis** [comment]
+1

**Comment by hgoona** [comment]
> @hgoona Yes, and yet it's special... All the CTRL+K and Chat functions use the model, but I haven't noticed any calls to ollama for the 4 Tab features (they do work though). And yet in my settings on the site, I have no calls to external APIs gpt-4o or other.

Wait - what?! Can I confirm: Does the Tab Prediction feature work even for Free users + Ollama ??

**Comment by takingurstuff** [comment]
+1

**Comment by tcsenpai** [comment]
Thanks, I was looking forward to something like this!
I will give it a try too, checking if autocomplete works as well.

**Comment by tcsenpai** [comment]
UPDATE: I  consistently get 403 errors both with LAN and public addresses. The curl command proposed by Cursor works in a terminal tho.

![image](https://github.com/user-attachments/assets/dce40b51-7d50-47e2-a896-405e334f27f7)

Also, this happens if I still try to execute a query:

![image](https://github.com/user-attachments/assets/73ebc286-4e66-4d2e-ac3e-bc7e1c9f0b85)

And I don't get any request to ollama.

Very weird.

**Comment by omutas** [comment]
@tcsenpai I had the same issue. Any solution?

**Comment by she11sh0cked** [comment]
Seems like a CORS issue. I never got it to work... Maybe you can disable web-security / CORS somehow?
![image](https://github.com/user-attachments/assets/9dc7d5ba-3568-42c0-96b4-4615572deaac)
![image](https://github.com/user-attachments/assets/573ca173-1ee3-4c2f-a3a1-cbc63f28d9e5)

**Comment by tcsenpai** [comment]
> @tcsenpai I had the same issue. Any solution?

Nope, I ended up trying another extension and the results were not fantastic, so I gave up completely until we get something like llama3.1 for coding.

**Comment by apalabrados** [comment]
@she11sh0cked Of course its CORS issue.
For make it working, you just need to enable CORS in the Ollama server.
Set the following environment variable:
`launchctl setenv OLLAMA_ORIGINS "*"`
and then run ollama server.

It works like a charm!

**Comment by hgoona** [comment]
@apalabrados does the Free tier of Cursor have access to the "Cursor (tab) Predictions" feature when using Ollama (plus your CORS solution) ??

**Comment by apalabrados** [comment]
> @apalabrados does the Free tier of Cursor have access to the "Cursor (tab) Predictions" feature when using Ollama LLMs (plus your CORS solution) ??

Yes! Attached an snapshot.
![Captura de pantalla 2024-10-04 a las 7 39 10](https://github.com/user-attachments/assets/f3e8102a-e8e8-4d07-96f2-b2fd6d506546)

**Comment by hgoona** [comment]
@apalabrados that's the "text prediction" that you've shown in your snapshot, right? What I'm talking about is the "Cursor Predictions" 

Looks like this where the "Tab" button appears in the view: 
![image](https://github.com/user-attachments/assets/adae73a3-77b2-490a-9ea6-9dba2ed76bb6)

Does that part also work?

**Comment by apalabrados** [comment]
@hgoona Hi again....
Yes, here you are the picture that shows it:
![Captura de pantalla 2024-10-04 a las 8 08 31](https://github.com/user-attachments/assets/a0acdaec-2a04-4eaf-bf21-991ac4b03073)

**Comment by hgoona** [comment]
@apalabrados thanks for confirming! Can i ask what Ollama models are you using for that? I'm not seeing that UI feature on mine...  Is it because I'm using groq models?🤔
![image](https://github.com/user-attachments/assets/ec06396b-576b-4965-a667-9221003ae8c4)

**Comment by apalabrados** [comment]
@hgoona The followings:
![Captura de pantalla 2024-10-04 a las 11 28 17](https://github.com/user-attachments/assets/fa1ddd7d-e9ea-409a-a204-b59da2abac1e)

Also, check if you have these options enabled:
![Captura de pantalla 2024-10-04 a las 11 29 28](https://github.com/user-attachments/assets/dd0f63a8-7f8d-42a5-be88-e418617373a3)

**Comment by grytsayev** [comment]
![image](https://github.com/user-attachments/assets/6b5f905b-3ecd-4a62-b4eb-c3064a8799b7)

Guys, I discovered the secret. Basically, for this to work, you need to trick the OpenAI key verification. So here are the steps:

1. **Activate** Turbo 3.5 GPT.
2. **Create** a real OpenAI API key.
3. **Paste** it into the OpenAI API Key section in Cursor settings.
4. **Reset** the base URL to default.
5. **Click** Verify and wait until it's verified successfully.
6. **Deactivate** all models except `llama3.1:latest`, for example.
7. **DON'T CHANGE THE API KEY**, just change the base URL and click Save.
8. Enjoy the working model!

**Note:** This feature may be patched soon, so if you want to continue using it, avoid updating!

**Comment by psmyrdek** [comment]
Just noticed that passing API key verification is no longer enough - seems like there's different handling of requests that prevents me from hitting local ollama.

<img width="506" alt="image" src="https://github.com/user-attachments/assets/cd16aed6-0f6b-413f-bef9-01e3dcd84165">

**Comment by srghma** [comment]
tried cursor + grok (llama-3.1-70b-versatile) - tab completitions didnt work (official name is https://api-docs.deepseek.com/guides/fim_completion I think)

**Comment by arthur-yh** [comment]
Can I not use Ngrok to expose api in network? I use localhost:11434, I can curl and test this llm api normally(in terminal). But it not work well in cursor, can anyone offer a solution? Well I can use Ngrok in company network...

local terminal post,work well 
`curl http://localhost:11434/v1/chat/completions -H "Content-Type: application/json" -H "Authorization: Bearer sk-abcdef1234567890ABCDEFGHIJKLMNOPQRSTUVWXz" -d '{
  "messages": [
    {
      "role": "system",
      "content": "You are a test assistant."
    },
    {
      "role": "user",
      "content": "Testing. Just say hi and nothing else."
    }
  ],
  "model": "codellama"
}'
{"id":"chatcmpl-432","object":"chat.completion","created":1731331175,"model":"codellama","system_fingerprint":"fp_ollama","choices":[{"index":0,"message":{"role":"assistant","content":"\nHi!"},"finish_reason":"stop"}],"usage":{"prompt_tokens":35,"completion_tokens":4,"total_tokens":39}}`

**Comment by marcusziade** [comment]
has anyone got the new Qwen models running locally in cursor? curious about performance on a 4090 or similar

**Comment by ozgrozer** [comment]
> has anyone got the new Qwen models running locally in cursor? curious about performance on a 4090 or similar

@marcusziade Yeah I tried the 32B on my MB M4 Max. It’s around the 12 tok/sec. [Here](https://x.com/ozgrozer/status/1856142448721793197)’s a video on X that I’m running it with LM Studio and Ngrok on Cursor.

**Comment by marcusziade** [comment]
> > has anyone got the new Qwen models running locally in cursor? curious about performance on a 4090 or similar
> 
> @marcusziade Yeah I tried the 32B on my MB M4 Max. It’s around the 12 tok/sec. [Here](https://x.com/ozgrozer/status/1856142448721793197)’s a video on X that I’m running it with LM Studio and Ngrok on Cursor.

yeah, I saw that. I'm gonna give LM Studio a try. I'm on arch with 7900xtx

**Comment by loktar00** [comment]
> Just noticed that passing API key verification is no longer enough - seems like there's different handling of requests that prevents me from hitting local ollama.
> 
> <img alt="image" width="506" src="https://private-user-images.githubusercontent.com/6187417/377942055-cd16aed6-0f6b-413f-bef9-01e3dcd84165.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0MzEwNjksIm5iZiI6MTczMTQzMDc2OSwicGF0aCI6Ii82MTg3NDE3LzM3Nzk0MjA1NS1jZDE2YWVkNi0wZjZiLTQxM2YtYmVmOS0wMWUzZGNkODQxNjUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTJUMTY1OTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDZkMDQ4YjQ1OWE3ZmVlNGY0MzZhZjUxYzdjMzlkNTM1ZDM1YTg2ZjFhYzA5NDY0Y2I1ZjIzZjhkYmU5NDc2MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.PZvoyedUoU43PkvmrVX-KnhpPwa06iIzFfR9I_9jjIg">

Getting this as well, I host on a different machine on my network as well so it's not due to "localhost" even :/

**Comment by loktar00** [comment]
https://github.com/user-attachments/assets/b692ad9f-e894-4150-b771-8e3a3d20007c

Runs really well on 2x3090's Used ngrok as well to get it working didn't need anything else.

**Comment by astr0gator** [comment]
Hi team, qwen 2.5-coder 32B is here and it's rad. 
Can anybody give hopes re implementing of ollama support? Frankly, it's game-changing/deal-breaking for many including me. I found [7 open issues](https://github.com/getcursor/cursor/issues?q=is%3Aissue+is%3Aopen+ollama) about implementing local llm in this repo

Thanks! 🙏🙏🙏

**Comment by loktar00** [comment]
> Hi team, qwen 2.5-coder 32B is here and it's rad. Can anybody give hopes re implementing of ollama support? Frankly, it's game-changing/deal-breaking for many including me. I found [7 open issues](https://github.com/getcursor/cursor/issues?q=is%3Aissue+is%3Aopen+ollama) about implementing local llm in this repo
> 
> Thanks! 🙏🙏🙏

It works fine, just use ngrok.

**Comment by ehartford** [comment]
> just use ngrok.

No. 

Cursor won't change until we make them change by hitting them in the wallet.

Be the change you want to see in the world - cancel and uninstall Cursor until they fix this intentional defect.

**Comment by usama-elm** [comment]
I think at this point we can see that cursor is not intent into opening up to local AI developpment.
Just as @ehartford shared, don't use it, take vanilla VSCode and install [Continue](https://www.continue.dev/)

**Comment by marcusziade** [comment]
is there any good neovim plugins that lets me connect to a private llm?

**Comment by arty-hlr** [comment]
> are there any good neovim plugins that let me connect to a private llm?

@marcusziade yes, vim-ai

**Comment by arty-hlr** [comment]
> Just noticed that passing API key verification is no longer enough - seems like there's different handling of requests that prevents me from hitting local ollama.
> <img alt="image" width="506" src="https://private-user-images.githubusercontent.com/6187417/377942055-cd16aed6-0f6b-413f-bef9-01e3dcd84165.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYyNzIwNzUsIm5iZiI6MTczNjI3MTc3NSwicGF0aCI6Ii82MTg3NDE3LzM3Nzk0MjA1NS1jZDE2YWVkNi0wZjZiLTQxM2YtYmVmOS0wMWUzZGNkODQxNjUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMDdUMTc0MjU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTM5MWNiMzZkYTI2OGM0MDYwNWFjZmIzMGNlMDViYzZhNmIxNDNkMTIzNDY1YTc0NWM3YjczMGEyMDE4MzVjOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.wjtJ_bk0KtvkEJuf3wLtC0HZdXr8ngLxGrLXnsNMiCk">

@psmyrdek Yes, seeing this as well, there is an issue for that error: https://github.com/getcursor/cursor/issues/2520

**Comment by pepo-ec** [comment]
I've tested using Ngrok and I could see that "curl" uses the POST method which is normal, while "cursor" is using the OPTIONS method and then the problems are triggered. How can I configure cursor to use POST as normally stated in the compatibility guide?

![image](https://github.com/user-attachments/assets/9c8f2759-257c-4cdb-8c18-f05b1bafbeaa)

**Comment by CyberT33N** [comment]
1. Allow Web Origins in Ollama. Will be needed for Cursor to send the request to ngrok to your ollama

```shell
sudo systemctl edit ollama.service

# Add this
[Service]
Environment="OLLAMA_ORIGINS=*"
```
- Make sure that you add the environment variable above this line `Edits below this comment will be discarded`

Then restart:
```shell
systemctl daemon-reload
systemctl restart ollama
```


<br><br>

2. Run your desired model by using ollama
- https://github.com/CyberT33N/ollama-cheat-sheet/blob/main/README.md#import-from-gguf
```shell
ollama run deepseek-v2-coder-lite
```


<br><br>

3. Tunneling your ollama rest api to cloud with ngrok
- https://github.com/ollama/ollama/blob/main/docs/faq.id opidmd#how-can-i-use-ollama-with-ngrok
```shell
ngrok http 11434 --host-header="localhost:11434"
```



<br><br>

4. Open Cursor > Cursor Settings > Models
  3.1 Disable all other models and add new model which has the same name as your hosted llm in ollama in my case `deepseek-v2-coder-lite`
  3.2 At section OpenAI API Key add your base url `https://xxxxxxxxxxxxx.ngrok-free.app/v1`. Add any random API Key And then click verify button
   - If you get 403 then something is not working with CORS.

<br><br>

5. Enjoy <3

**Comment by Justin-12138** [comment]
@CyberT33N I ran it well yesterday，But It doesn't work anymore

![Image](https://github.com/user-attachments/assets/14f231aa-d9a1-4ef7-89a0-99c9887e9915)

**Comment by CyberT33N** [comment]
> [@CyberT33N](https://github.com/CyberT33N) I ran it well yesterday，But It doesn't work anymore
> 
> ![Image](https://github.com/user-attachments/assets/14f231aa-d9a1-4ef7-89a0-99c9887e9915)

For me the sulotion posted above does anyway only work with the Chat mode and composer normal mode. When using Composer agent it says anyway that only gpt-40 and claude-sonnet-3.5 are supported. Since then I was not deeping further into it but maybe disabling all other models in the settings area and then rename your custom model like the expected once. However, this whole process of getting local llms runnign with cursor ist just a pain. It should be supported out of the box by cursor.

**Comment by alexellis** [comment]
I got this working with [inlets.dev](https://inlets.dev) to expose a stable domain/URL over HTTPS. I had to run a tiny Go proxy to change the Host: header - that is the main issue people see with `OPTIONS` failing from Cursor due to cors, but working from curl.

I also override the models endpoint so that cursor thinks it's working with gpt-4o-mini. You can run `cp llama3.2 gpt-4o-mini` go rename a model in Ollama.

I wasn't able to fully save the custom URL in cursor until I added credit to OpenAI's API and used an API key, free-tier keys do not work and won't be validated.

The caveat is that it only works for chat, edits, explanations - it will not work with tab completion, because cursor have hard-coded the URL. You'd have to use a TLS-intercepting MITM proxy to intercept the traffic and send it to ollama instead.

<img width="1470" alt="Image" src="https://github.com/user-attachments/assets/d592d11f-8cbc-4862-89db-cb07267ec5a6" />

```golang
package main

import (
	"log"
	"net/http"
	"net/http/httputil"
	"net/url"
)

func main() {
	// Parse the upstream URL
	target, err := url.Parse("http://127.0.0.1:11434")
	if err != nil {
		log.Fatalf("Failed to parse target URL: %v", err)
	}

	// Create the reverse proxy
	proxy := httputil.NewSingleHostReverseProxy(target)

	// Modify the request before sending it upstream
	proxy.Director = func(req *http.Request) {
		//		req.Header.Set("Origin", "http://127.0.0.1")
		req.Header.Del("X-Forwarded-Host")
		req.Header.Del("X-Forwarded-For")

		req.URL.Scheme = target.Scheme
		req.URL.Host = target.Host
		req.Host = "localhost" // Set the Host header to "localhost"

		for k, v := range req.Header {
			log.Printf("%s=%v", k, v)
		}
		log.Printf("%s", req.URL.String())
		log.Println()

	}

	mux := http.NewServeMux()

	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization")

		if r.URL.Path == "/v1/models" {
			log.Println("Model list requested")
			w.Header().Set("Content-Type", "application/json")
			w.Write([]byte(`{
  "object": "list",
  "data": [
    {
      "id": "gpt-4o-mini",
      "object": "model",
      "created": 1740000216,
      "owned_by": "openai"
    }
  ]
}`))
			return
		}

		log.Println("Request received")
		proxy.ServeHTTP(w, r)
	})

	// Start the server
	log.Println("Starting proxy server on :8080")
	if err := http.ListenAndServe(":8080", mux); err != nil {
		log.Fatalf("Failed to start server: %v", err)
	}
}
```

**Comment by littlebirdxu** [comment]
How Can I switch composer to agent mode? I can't find it.

**Comment by manupatet** [comment]
Works for me with simple SSH (port forwarding). I ran into three issues trying to get it to work with local ollama, but got a green in the end.

- Ollama refusing to serve anything other than `localhost`: Simple fix (add `export OLLAMA_HOST=0.0.0.0`)
- Ollama running into CORS issue and throwing 403 when accessed via a URL: simple fix ( add `export OLLAMA_ORIGINS='*'`)
- Validate CORS issue with this command `curl -X GET http://localhost:11434 -H "Origin: http://example.com" -H "Access-Control-Request-Method: GET" -I`, you should see `HTTP 200 OK`, otherwise there's still issue.
- Run this command : `ollama cp <your_model_name> default`. This is because while verifying Cursor looks for `default` for verifying api key.
- Run this command `ssh -R 80:localhost:11434 serveo.net`. It'll put out a URL 
```
$ ssh -R 80:localhost:11434 serveo.net
Forwarding HTTP traffic from https://95af4527fcb2f8c69f262c8aff244a6a.serveo.net
```
- Over at Cursor-ai settings tab (the cog), click 'Models' and add a new model (one from `ollama list`) and save.
  -- In section `OpenAI API Key`, click on `override OpenAI Base URL` and enter the above URL (append a `/v1` as shown), and hit `save`.
```
https://93afe527fcb2f8c69f262c8aff244a6a.serveo.net/v1
```
- Put something random in API Key field and hit `Verify`
- Should turn green.

Cursor-ai version: `0.46.11`
Ollama version: 0.5.13
Mac OS: Sequoia 15.3

**Comment by arthurmf** [comment]
> Works for me with simple SSH (port forwarding). I ran into three issues trying to get it to work with local ollama, but got a green in the end.
> 
> * Ollama refusing to serve anything other than `localhost`: Simple fix (add `export OLLAMA_HOST=0.0.0.0`)
> * Ollama running into CORS issue and throwing 403 when accessed via a URL: simple fix ( add `export OLLAMA_ORIGINS='*'`)
> * Validate CORS issue with this command `curl -X GET http://localhost:11434 -H "Origin: http://example.com" -H "Access-Control-Request-Method: GET" -I`, you should see `HTTP 200 OK`, otherwise there's still issue.
> * Run this command : `ollama cp <your_model_name> default`. This is because while verifying Cursor looks for `default` for verifying api key.
> * Run this command `ssh -R 80:localhost:11434 serveo.net`. It'll put out a URL
> 
> ```
> $ ssh -R 80:localhost:11434 serveo.net
> Forwarding HTTP traffic from https://95af4527fcb2f8c69f262c8aff244a6a.serveo.net
> ```
> 
> * Over at Cursor-ai settings tab (the cog), click 'Models' and add a new model (one from `ollama list`) and save.
>   -- In section `OpenAI API Key`, click on `override OpenAI Base URL` and enter the above URL (append a `/v1` as shown), and hit `save`.
> 
> ```
> https://93afe527fcb2f8c69f262c8aff244a6a.serveo.net/v1
> ```
> 
> * Put something random in API Key field and hit `Verify`
> * Should turn green.
> 
> Cursor-ai version: `0.46.11` Ollama version: 0.5.13 Mac OS: Sequoia 15.3

@manupatet how do you add new model? It seems like my cursor doesn't allow me to add new models..
1-I click in on "+ Add Model"
2- The text box opens and I type the model name: "qwen2.5-coder:32b"
3- I click on "+ Add Model" and nothing happens.

**Comment by sb52pj** [comment]
> Works for me with simple SSH (port forwarding). I ran into three issues trying to get it to work with local ollama, but got a green in the end.
> 
> * Ollama refusing to serve anything other than `localhost`: Simple fix (add `export OLLAMA_HOST=0.0.0.0`)
> * Ollama running into CORS issue and throwing 403 when accessed via a URL: simple fix ( add `export OLLAMA_ORIGINS='*'`)
> * Validate CORS issue with this command `curl -X GET http://localhost:11434 -H "Origin: http://example.com" -H "Access-Control-Request-Method: GET" -I`, you should see `HTTP 200 OK`, otherwise there's still issue.
> * Run this command : `ollama cp <your_model_name> default`. This is because while verifying Cursor looks for `default` for verifying api key.
> * Run this command `ssh -R 80:localhost:11434 serveo.net`. It'll put out a URL
> 
> ```
> $ ssh -R 80:localhost:11434 serveo.net
> Forwarding HTTP traffic from https://95af4527fcb2f8c69f262c8aff244a6a.serveo.net
> ```
> 
> * Over at Cursor-ai settings tab (the cog), click 'Models' and add a new model (one from `ollama list`) and save.
>   -- In section `OpenAI API Key`, click on `override OpenAI Base URL` and enter the above URL (append a `/v1` as shown), and hit `save`.
> 
> ```
> https://93afe527fcb2f8c69f262c8aff244a6a.serveo.net/v1
> ```
> 
> * Put something random in API Key field and hit `Verify`
> * Should turn green.
> 
> Cursor-ai version: `0.46.11` Ollama version: 0.5.13 Mac OS: Sequoia 15.3

superb! thanx!!!!

**Comment by manupatet** [comment]
> [@manupatet](https://github.com/manupatet) how do you add new model? It seems like my cursor doesn't allow me to add new models.. 
>1-I click in on "+ Add Model" 
>2- The text box opens and I type the model name: "qwen2.5-coder:32b" 
>3- I click on "+ Add Model" and nothing happens.

Just hit enter after step 2 @arthurmf, i.e. skip your step 3 (its a UI issue)

**Comment by manupatet** [comment]
> but fail frequently on long requests.

@sb52pj for me it times out after a minute, so an option is to try a quantized version of your model

**Comment by kcolemangt** [comment]
You can use [llm-router](https://github.com/kcolemangt/llm-router/releases) to add Groq and Ollama providers. v0.1.0 adds support for cursor’s verification checks.

**Comment by mavericreal** [comment]
> Works for me with simple SSH (port forwarding). I ran into three issues trying to get it to work with local ollama, but got a green in the end.
> 
> * Ollama refusing to serve anything other than `localhost`: Simple fix (add `export OLLAMA_HOST=0.0.0.0`)
> * Ollama running into CORS issue and throwing 403 when accessed via a URL: simple fix ( add `export OLLAMA_ORIGINS='*'`)
> * Validate CORS issue with this command `curl -X GET http://localhost:11434 -H "Origin: http://example.com" -H "Access-Control-Request-Method: GET" -I`, you should see `HTTP 200 OK`, otherwise there's still issue.
> * Run this command : `ollama cp <your_model_name> default`. This is because while verifying Cursor looks for `default` for verifying api key.
> * Run this command `ssh -R 80:localhost:11434 serveo.net`. It'll put out a URL
> 
> ```
> $ ssh -R 80:localhost:11434 serveo.net
> Forwarding HTTP traffic from https://95af4527fcb2f8c69f262c8aff244a6a.serveo.net
> ```
> 
> * Over at Cursor-ai settings tab (the cog), click 'Models' and add a new model (one from `ollama list`) and save.
>   -- In section `OpenAI API Key`, click on `override OpenAI Base URL` and enter the above URL (append a `/v1` as shown), and hit `save`.
> 
> ```
> https://93afe527fcb2f8c69f262c8aff244a6a.serveo.net/v1
> ```
> 
> * Put something random in API Key field and hit `Verify`
> * Should turn green.
> 
> Cursor-ai version: `0.46.11` Ollama version: 0.5.13 Mac OS: Sequoia 15.3

This worked for me too. The only difference is that I had to run:

`ollama cp <your_model_name> gpt-4o`

Instead of copying it to default. 

Thank you @manupatet

**Comment by arthurwolf** [comment]
I would probably not be a lot of extra coding (especially if you use Cursor ... ehehe) for you guys to add native support for ollama / locally run models, and we get better local models every week at this point, they are increasingly becoming a viable option (see the release of Qwen3 today...).

I would really like a less convoluted/more streamlined way of using local models, it sorts of sucks that I have to lose the ability to use openai models if I want to use ollama models...

Thanks :)

**Comment by minkooseo** [comment]
> Works for me with simple SSH (port forwarding). I ran into three issues trying to get it to work with local ollama, but got a green in the end.
> 
> * Ollama refusing to serve anything other than `localhost`: Simple fix (add `export OLLAMA_HOST=0.0.0.0`)
> * Ollama running into CORS issue and throwing 403 when accessed via a URL: simple fix ( add `export OLLAMA_ORIGINS='*'`)
> * Validate CORS issue with this command `curl -X GET http://localhost:11434 -H "Origin: http://example.com" -H "Access-Control-Request-Method: GET" -I`, you should see `HTTP 200 OK`, otherwise there's still issue.
> * Run this command : `ollama cp <your_model_name> default`. This is because while verifying Cursor looks for `default` for verifying api key.
> * Run this command `ssh -R 80:localhost:11434 serveo.net`. It'll put out a URL
> 
> ```
> $ ssh -R 80:localhost:11434 serveo.net
> Forwarding HTTP traffic from https://95af4527fcb2f8c69f262c8aff244a6a.serveo.net
> ```
> 
> * Over at Cursor-ai settings tab (the cog), click 'Models' and add a new model (one from `ollama list`) and save.
>   -- In section `OpenAI API Key`, click on `override OpenAI Base URL` and enter the above URL (append a `/v1` as shown), and hit `save`.
> 
> ```
> https://93afe527fcb2f8c69f262c8aff244a6a.serveo.net/v1
> ```
> 
> * Put something random in API Key field and hit `Verify`
> * Should turn green.
> 
> Cursor-ai version: `0.46.11` Ollama version: 0.5.13 Mac OS: Sequoia 15.3

Works perfectly. For lm studio, CORS can be set using settings. There's no required corresponding step to OLLAMA_HOST=0.0.0.0.

<img width="523" alt="Image" src="https://github.com/user-attachments/assets/b861d91c-85f7-4dba-9027-d926a5130fd9" />

**Comment by henry2man** [comment]
With this setup, is Agent mode working with any model? If so, what are the recommended models?
